packages/core/src/actions/smartAccount/internal/runMiddlewareStack.ts

```diff
@@ -22,38 +24,52 @@ const asyncPipe =
     return result;
   };
 
-export const _runMiddlewareStack: <
+export async function _runMiddlewareStack<
   TTransport extends Transport = Transport,
   TChain extends Chain | undefined = Chain | undefined,
   TAccount extends SmartContractAccount | undefined =
     | SmartContractAccount
     | undefined,
   TContext extends Record<string, any> | undefined =
     | Record<string, any>
-    | undefined
+    | undefined,
+  TEntryPointVersion extends GetEntryPointFromAccount<TAccount> = GetEntryPointFromAccount<TAccount>
 >(
   client: BaseSmartAccountClient<TTransport, TChain, TAccount>,
   args: {
-    uo: Deferrable<UserOperationStruct>;
-    overrides?: UserOperationOverrides;
+    uo: Deferrable<UserOperationStruct<TEntryPointVersion>>;
     context?: TContext;
-  } & GetAccountParameter<TAccount>
-) => Promise<UserOperationStruct> = async (client, args) => {
+  } & GetAccountParameter<TAccount> &
+    UserOperationOverridesParameter<TEntryPointVersion>
+): Promise<UserOperationStruct<TEntryPointVersion>> {
   const { uo, overrides, account = client.account, context } = args;
   if (!account) {
     throw new AccountNotFoundError();
   }
+
+  const entryPoint = account.getEntryPoint();
+
+  const overridePaymasterData =
+    overrides &&
+    ((entryPoint.version === "0.6.0" &&
+      (overrides as UserOperationOverrides<"0.6.0">).paymasterAndData !=
+        null) ||
+      (entryPoint.version === "0.7.0" &&
+        (overrides as UserOperationOverrides<"0.7.0">).paymasterData != null));
+
   const result = await asyncPipe(
     client.middleware.dummyPaymasterAndData,
     client.middleware.feeEstimator,
     client.middleware.gasEstimator,
     client.middleware.customMiddleware,
-    overrides?.paymasterAndData
+    overridePaymasterData
       ? overridePaymasterDataMiddleware
       : client.middleware.paymasterAndData,
     client.middleware.userOperationSimulator,
     client.middleware.signUserOperation
   )(uo, { overrides, feeOptions: client.feeOptions, account, client, context });
 
-  return resolveProperties<UserOperationStruct>(result);
-};
+  return resolveProperties(result) as Promise<
+    UserOperationStruct<GetEntryPointFromAccount<TAccount>>
+  >;
```

Comment:
typecasting should be avoided when it is possible
```suggestion
  return resolveProperties<
    UserOperationStruct<GetEntryPointFromAccount<TAccount>>
  >(result);
```